src/uucore/src/lib/macros.rs

```diff
@@ -5,31 +5,82 @@
 // For the full copyright and license information, please view the LICENSE
 // file that was distributed with this source code.
 
-/// Deduce the name of the binary from the current source code filename.
-///
-/// e.g.: `src/uu/cp/src/cp.rs` -> `cp`
+/// Get the executable path (as `OsString`).
+#[macro_export]
+macro_rules! executable_os(
+    () => ({
+        &uucore::args_os().next().unwrap()
+    })
+);
+
+/// Get the executable path (as `String`; lossless).
 #[macro_export]
 macro_rules! executable(
     () => ({
-        let module = module_path!();
-        let module = module.split("::").next().unwrap_or(module);
-        if &module[0..3] == "uu_" {
-            &module[3..]
+        let exe = match $crate::executable_os!().to_str() {
+            // * UTF-8
+            Some(s) => s.to_string(),
+            // * "lossless" debug format (if/when `executable_os!()` is not well-formed UTF-8)
+            None => format!("{:?}", $crate::executable_os!())
+        };
+        &exe.to_owned()
+    })
+);
+
+/// Get the executable name.
+#[macro_export]
+macro_rules! executable_name(
+    () => ({
+        let stem = &std::path::Path::new($crate::executable_os!()).file_stem().unwrap().to_owned();
+        let exe = match stem.to_str() {
+            // * UTF-8
+            Some(s) => s.to_string(),
+            // * "lossless" debug format (if/when `executable_os!()` is not well-formed UTF-8)
+            None => format!("{:?}", stem)
+        };
+        &exe.to_owned()
+    })
+);
+
+/// Derive the utility name.
+#[macro_export]
+macro_rules! util_name(
+    () => ({
+        let crate_name = env!("CARGO_PKG_NAME");
+        let name = if crate_name.starts_with("uu_") {
+            &crate_name[3..]
         } else {
-            module
-        }
+            crate_name
+        };
+        &name.to_owned()
     })
 );
 
+//====
+
+/// Derive the complete execution phrase for "usage".
+#[macro_export]
+macro_rules! execution_phrase(
+    () => ({
+        let exe = if (executable_name!() == util_name!()) {
```

Comment:
Hmm, but what if you rename the binary?
For example, if you rename `sort` to `uu_sort`:
`cargo b -p uu_sort && cp target/debug/sort uu_sort && ./uu_sort --help` will print `USAGE:
    ./uu_sort sort`.


FWIW, if you install uutils from the official arch repos the binaries will be installed with exactly that name (`uu_*`)

Reply:
Yes, I meant to put a `ToDO` there to think further about the best way to deal with busybox-type binary vs single util binaries and renames for each type (including possible prefixes for busybox-type).

Reply:
Maybe we could set a global flag when we are invoked with the busybox binary?

Reply:
Yes, I was thinking we'd need some global signal as well; say defaulting to single-call, and set to multi-call by a busybox binary. I'm trying to think of a way to build this into uucore (correctly 😄) and simultaneously simplify the `coreutils` multi-call start up logic.